docs/tutorial/write-a-machine-charm-for-tinyproxy.md

```diff
@@ -0,0 +1,577 @@
+
+```text
+charmcraft pack
+```
+
+This created the charm:
+
+```text
+CONTRIBUTING.md  charmcraft.yaml  src                    tox.ini
+LICENSE          lib              tests                  uv.lock
+README.md        pyproject.toml   tinyproxy_amd64.charm
+```
+
+(The charm filename depends on architecture)
+
+### Deploy your charm
+
+TODO: Add commentary to this section.
+
+Open another terminal, using `multipass shell juju-sandbox`. Use this terminal to watch Juju status`. In the other terminal, deploy the charm:
+
+```text
+juju deploy ./tinyproxy_amd64.charm
+```
+
+(The charm filename depends on architecture)
+
+When the charm is ready, you'll see:
+
+```text
+Model    Controller     Cloud/Region         Version  SLA          Timestamp
+testing  concierge-lxd  localhost/localhost  3.6.8    unsupported  09:00:38+08:00
+
+App        Version  Status  Scale  Charm      Channel  Rev  Exposed  Message
+tinyproxy           active      1  tinyproxy             0  no
+
+Unit          Workload  Agent  Machine  Public address  Ports  Message
+tinyproxy/0*  active    idle   0        10.71.67.208
+
+Machine  State    Address       Inst id        Base          AZ  Message
+0        started  10.71.67.208  juju-8e7bd9-0  ubuntu@22.04      Running
+```
+
+Make a note of the IP address of the machine (10.71.67.208).
+
+### Try the proxy
+
+TODO: Add commentary to this section.
+
+```text
+curl <address>:8000/example/
+```
+
+Output:
+
+```text
+<!doctype html>
+<html>
+<head>
+    <title>Example Domain</title>
+    ...
+</head>
+
+<body>
+<div>
+    <h1>Example Domain</h1>
+    <p>This domain is for use in illustrative examples in documents. You may use this
+    domain in literature without prior coordination or asking for permission.</p>
+    <p><a href="https://www.iana.org/domains/example">More information...</a></p>
+</div>
+</body>
+</html>
+```
+
+### Change the configuration
+
+TODO: Add commentary to this section.
+
+```text
+juju config tinyproxy slug=foo
+curl <address>:8000/foo/  # Returns the same HTML as before.
+curl <address>:8000/example/  # Returns an error page.
+juju config tinyproxy slug=foo/bar  # Puts the charm into blocked status.
+juju config tinyproxy --reset slug  # Makes the charm active, with the original config.
+```
+
+## Write unit tests for your charm
+
+### Write tests for the helper module
+
+TODO: Add commentary to this section.
+
+`tests/unit/test_tinyproxy.py`:
+
+```python
+import pytest
+
+from charm import tinyproxy
+
+
+def test_slug_valid():
+    tinyproxy.check_slug("example")  # No error raised.
+
+
+def test_slug_invalid():
+    with pytest.raises(ValueError):
+        tinyproxy.check_slug("foo/bar")
+```
+
+### Write state-transition tests
+
+TODO: Add commentary to this section.
+
+`tests/unit/test_charm.py`:
+
+```python
+from dataclasses import dataclass
+
+import pytest
+from ops import testing
+
+from charm import PORT, TinyproxyCharm
+
+
+@dataclass
+class MockTinyproxy:
```

Comment:
It seems a bit odd to use a dataclass here. You're only really using it for the init, right? Not the rest of what a dataclass gives, like eq, hash, easy dumping to JSON, etc. I feel like this would be more natural as just a plain class with an explicit init.

I do, however, like how clean having the workload interaction in a module makes this mocking.